### ISSUE TITLE
`isinstance` doesn't narrow when using union

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#13438]
```py
a: int | str
assert isinstance(a, int | None)
reveal_type(a)
```
[playground](https://mypy-play.net/?mypy=latest&python=3.10&gist=4850ebb29a23c6039769a0b05493e3aa)

**Comment by JelleZijlstra** [comment]
This seems specific to `| None`. If you do `assert isinstance(a, int | list)` instead, it correctly narrows to `int`.

**Comment by SakshiSingh25** [comment]
Hi, I'm interested in working on this issue. Kindly, guide me with the procedure.

**Comment by KotlinIsland** [comment]
@SakshiSingh25 Start here: https://github.com/python/mypy/blob/master/CONTRIBUTING.md

This issue is likely due to `strict_optional` in some way, so search for that.

**Comment by brianschubert** [comment]
This was fixed in #17415
